File: tests/UnitTestApp/FunctionalTests/ANInstreamVideoAdTestCase.m
/*
 *
 *    Copyright 2017 APPNEXUS INC
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */


#import <XCTest/XCTest.h>

#import "ANGlobal.h"
#import "ANTestGlobal.h"
#import "ANHTTPStubbingManager.h"
#import "ANSDKSettings+PrivateMethods.h"

#import "ANInstreamVideoAd.h"
#import "ANInstreamVideoAd+Test.h"
#import "ANVideoAdPlayer.h"
#import "XandrAd.h"


static NSString   *placementID      = @"12534678";
static NSInteger   memberID         = 958;
static NSString   *inventoryCode    = @"trucksmash";



@interface ANInstreamVideoAdTestCase : XCTestCase <ANInstreamVideoAdLoadDelegate>

@property (nonatomic, readwrite, strong)  ANInstreamVideoAd  *instreamVideoAd;

@property (nonatomic, strong)  XCTestExpectation  *expectationLoadVideoAd;

@property (nonatomic, strong)  NSURLRequest  *request;
@property (nonatomic, strong)  NSDictionary  *jsonRequestBody;

@end




@implementation ANInstreamVideoAdTestCase

#pragma mark - Test lifecycle.

- (void)setUp
{
    [super setUp];
    [[ANSDKSettings sharedInstance] optionalSDKInitialization:nil];
    // Init here if not the tests will crash
    [[XandrAd sharedInstance] initWithMemberID:1 preCacheRequestObjects:true completionHandler:nil];
    [[ANHTTPStubbingManager sharedStubbingManager] enable];
    [ANHTTPStubbingManager sharedStubbingManager].ignoreUnstubbedRequests = YES;

    [ANHTTPStubbingManager sharedStubbingManager].broadcastRequests = YES;
    [[NSNotificationCenter defaultCenter] addObserver: self
                                             selector: @selector(requestLoaded:)
                                                 name: kANHTTPStubURLProtocolRequestDidLoadNotification
                                               object: nil ];

    self.request = nil;
}

- (void)tearDown
{
    [super tearDown];
    [ANHTTPStubbingManager sharedStubbingManager].broadcastRequests = NO;
    [[ANHTTPStubbingManager sharedStubbingManager] removeAllStubs];
    [[ANHTTPStubbingManager sharedStubbingManager] disable];

    self.instreamVideoAd = nil;
    self.expectationLoadVideoAd = nil;
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    for (UIView *additionalView in [[ANGlobal getKeyWindow].rootViewController.view subviews]){
        [additionalView removeFromSuperview];
    }
}




#pragma mark - Test methods.

- (void)testInitializeWithPlacementID
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];

    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];

    [instreamVideoAd loadAdWithDelegate:self];

    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];

    //
    XCTAssertEqual(instreamVideoAd.placementId, placementID);
    XCTAssertEqual([self.jsonRequestBody[@"tags"][0][@"id"] integerValue], [placementID integerValue]);

}

- (void)testInitializeWithForceCreativeID
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    instreamVideoAd.forceCreativeId = 135482485;
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];

    [instreamVideoAd loadAdWithDelegate:self];

    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];

    //
    XCTAssertEqual(135482485, instreamVideoAd.forceCreativeId);
    XCTAssertEqual([self.jsonRequestBody[@"tags"][0][@"force_creative_id"] integerValue], 135482485);

}

- (void)testInitializeWithMemberIDAndCode
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithMemberId:958 inventoryCode:@"trucksmash"];

    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];

    [instreamVideoAd loadAdWithDelegate:self];

    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];

    XCTAssertEqual(instreamVideoAd.memberId, memberID);
    XCTAssertEqual(instreamVideoAd.inventoryCode, inventoryCode);
    XCTAssertEqual([self.jsonRequestBody[@"member_id"] integerValue], memberID);
    XCTAssertTrue([self.jsonRequestBody[@"tags"][0][@"code"] isEqualToString:inventoryCode]);
}



- (void)testCustomKeywordsAdded
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    [instreamVideoAd addCustomKeywordWithKey:@"force_creative_id" value:@"123456789"];
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    XCTAssertEqualObjects(self.jsonRequestBody[@"tags"][0][@"keywords"][0][@"key"], @"force_creative_id");
    XCTAssertEqualObjects(self.jsonRequestBody[@"tags"][0][@"keywords"][0][@"value"][0], @"123456789");

    XCTAssertEqual(instreamVideoAd.placementId, placementID);
}

- (void)testTrafficSourceCode
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    [instreamVideoAd setTrafficSourceCode:@"Xandr"];
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    XCTAssertEqualObjects(self.jsonRequestBody[@"tags"][0][@"traffic_source_code"], @"Xandr");
    XCTAssertNil(self.jsonRequestBody[@"tags"][0][@"ext_inv_code"]);
    instreamVideoAd = nil;
}


- (void)testExtInvCode
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    [instreamVideoAd setExtInvCode:@"Xandr"];
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    XCTAssertEqualObjects(self.jsonRequestBody[@"tags"][0][@"ext_inv_code"], @"Xandr");
    XCTAssertNil(self.jsonRequestBody[@"tags"][0][@"traffic_source_code"]);
    instreamVideoAd = nil;
}

- (void)testTrafficSourceCodeAndExtInvCodeSet
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];
    [instreamVideoAd setExtInvCode:@"Xandr-ext_inv_code"];
    [instreamVideoAd setTrafficSourceCode:@"Xandr-traffic_source_code"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    XCTAssertEqualObjects(self.jsonRequestBody[@"tags"][0][@"traffic_source_code"], @"Xandr-traffic_source_code");
    XCTAssertEqualObjects(self.jsonRequestBody[@"tags"][0][@"ext_inv_code"], @"Xandr-ext_inv_code");
    instreamVideoAd = nil;
}

- (void)testTrafficSourceCodeAndExtInvCodeNotSet
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    NSDictionary *tag = self.jsonRequestBody[@"tags"][0];
    XCTAssertFalse([tag objectForKey:@"ext_inv_code"]);
    XCTAssertFalse([tag objectForKey:@"traffic_source_code"]);
    instreamVideoAd = nil;
}

- (void)testTrafficSourceCodeAndExtInvCodeNullSet
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];
    [instreamVideoAd setExtInvCode:NULL];
    [instreamVideoAd setTrafficSourceCode:NULL];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    NSDictionary *tag = self.jsonRequestBody[@"tags"][0];
    XCTAssertFalse([tag objectForKey:@"ext_inv_code"]);
    XCTAssertFalse([tag objectForKey:@"traffic_source_code"]);
    instreamVideoAd = nil;
}


- (void)testAdDuration {
        [self initializeInstreamVideoWithAllProperties];
        NSLog(@"reached here");
        XCTAssertNotNil(self.instreamVideoAd);
        NSUInteger duration = [self.instreamVideoAd getAdDuration];
        XCTAssertNotEqual(duration, 0);
 }

-(void) testVastCreativeURL {
    
        [self initializeInstreamVideoWithAllProperties];
        NSLog(@"reached here");
        XCTAssertNotNil(self.instreamVideoAd);
        NSString *vastcreativeTag = [self.instreamVideoAd getVastURL];
        XCTAssertNotNil(vastcreativeTag);
        XCTAssertNotEqual(vastcreativeTag.length, 0);
        XCTAssertNotNil(vastcreativeTag);
        XCTAssertEqual(vastcreativeTag, @"http://sampletag.com");
}

-(void) testVastCreativeXML {
    
    [self initializeInstreamVideoWithAllProperties];
    NSLog(@"reached here");
    XCTAssertNotNil(self.instreamVideoAd);
    NSString *vastcreativeXMLTag = [self.instreamVideoAd getVastXML];
    XCTAssertNotNil(vastcreativeXMLTag);
    XCTAssertNotEqual(vastcreativeXMLTag.length, 0);
    XCTAssertNotNil(vastcreativeXMLTag);
    XCTAssertEqual(vastcreativeXMLTag, @"http://sampletag.com");
}

-(void) testCreativeTag {
        [self initializeInstreamVideoWithAllProperties];
        NSLog(@"reached here");
        XCTAssertNotNil(self.instreamVideoAd);
        NSString *creativeTag = [self.instreamVideoAd getCreativeURL];
        XCTAssertNotEqual(creativeTag.length, 0);
        XCTAssertNotNil(creativeTag);
        XCTAssertEqual(creativeTag, @"http://sampletag.com");
}

-(void) testAdDurationNotSet {
        [self initializeInstreamVideoWithNoProperties];
        XCTAssertNotNil(self.instreamVideoAd);
        NSUInteger duration = [self.instreamVideoAd getAdDuration];
        XCTAssertEqual(duration, 0);
}

-(void) testVastCreativeValuesNotSet {
        [self initializeInstreamVideoWithNoProperties];
        XCTAssertNotNil(self.instreamVideoAd);
        NSString *vastcreativeTag = [self.instreamVideoAd getVastURL];
        XCTAssertEqual(vastcreativeTag.length, 0);
}

-(void) testCreativeValuesNotSet {
    
        [self initializeInstreamVideoWithNoProperties];
        XCTAssertNotNil(self.instreamVideoAd);
        NSString *creativeTag = [self.instreamVideoAd getCreativeURL];
        XCTAssertEqual(creativeTag.length, 0);
}

-(void) testVastCreativeXMLValuesNotSet {
    [self initializeInstreamVideoWithNoProperties];
    XCTAssertNotNil(self.instreamVideoAd);
    NSString *vastcreativeXMLTag = [self.instreamVideoAd getVastXML];
    XCTAssertEqual(vastcreativeXMLTag.length, 0);
}

-(void) testPlayHeadTimeForVideoSet {
    [self initializeInstreamVideoWithNoProperties];
    XCTAssertNotNil(self.instreamVideoAd);
    NSUInteger duration = [self.instreamVideoAd getAdPlayElapsedTime];
    XCTAssertNotEqual(duration, 0);
}

- (void)testVerticalVideoOrientation
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    
    [self stubRequestWithResponse:@"SuccessfulVerticalVideoAdResponse"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    
    //
    XCTAssertEqual(instreamVideoAd.getVideoOrientation, ANPortrait);
    
}


- (void)testLandscapeVideoOrientation
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    
    [self stubRequestWithResponse:@"SuccessfulInstreamVideoAdResponse"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    
    //
    XCTAssertEqual(instreamVideoAd.getVideoOrientation, ANLandscape);
    
}



- (void)testSquareVideoOrientation
{
    ANInstreamVideoAd  *instreamVideoAd  = [[ANInstreamVideoAd alloc] initWithPlacementId:placementID];
    
    [self stubRequestWithResponse:@"SuccessfulSquareInstreamVideoAdResponse"];
    
    [instreamVideoAd loadAdWithDelegate:self];
    
    self.expectationLoadVideoAd = [self expectationWithDescription:[NSString stringWithFormat:@"%s", __PRETTY_FUNCTION__]];
    [self waitForExpectationsWithTimeout:kAppNexusRequestTimeoutInterval handler:nil];
    
    //
    XCTAssertEqual(instreamVideoAd.getVideoOrientation, ANSquare);
    
}

#pragma mark - Helper methods.

-(void) initializeInstreamVideoWithAllProperties {
    self.instreamVideoAd = [[ANInstreamVideoAd alloc] init];
    self.instreamVideoAd.adPlayer = [[ANVideoAdPlayer alloc] init];
    self.instreamVideoAd.adPlayer.videoDuration = 10;
    self.instreamVideoAd.adPlayer.creativeURL = @"http://sampletag.com";
    self.instreamVideoAd.adPlayer.vastURLContent = @"http://sampletag.com";
    self.instreamVideoAd.adPlayer.vastXMLContent = @"http://sampletag.com";
    
}

-(void) initializeInstreamVideoWithNoProperties {
    self.instreamVideoAd = [[ANInstreamVideoAd alloc] init];
    self.instreamVideoAd.adPlayer = [[ANVideoAdPlayer alloc] init];
}

- (void) stubRequestWithResponse:(NSString *)responseName
{
    NSBundle *currentBundle = [NSBundle bundleForClass:[self class]];
    NSString *baseResponse = [NSString stringWithContentsOfFile: [currentBundle pathForResource:responseName
                                                                                         ofType:@"json" ]
                                                       encoding: NSUTF8StringEncoding
                                                          error: nil ];

    ANURLConnectionStub  *requestStub  = [[ANURLConnectionStub alloc] init];

    requestStub.requestURL    = [[[ANSDKSettings sharedInstance] baseUrlConfig] utAdRequestBaseUrl];
    requestStub.responseCode  = 200;
    requestStub.responseBody  = baseResponse;

    [[ANHTTPStubbingManager sharedStubbingManager] addStub:requestStub];
}

- (void)requestLoaded:(NSNotification *)notification
{
    NSURLRequest  *incomingRequest  = notification.userInfo[kANHTTPStubURLProtocolRequest];

    NSString  *requestString  = [[incomingRequest URL] absoluteString];
    NSString  *searchString   = [[[ANSDKSettings sharedInstance] baseUrlConfig] utAdRequestBaseUrl];

    if (!self.request && [requestString containsString:searchString])
    {
        self.request          = notification.userInfo[kANHTTPStubURLProtocolRequest];
        self.jsonRequestBody  = [ANHTTPStubbingManager jsonBodyOfURLRequestAsDictionary:self.request];
    }
}




#pragma mark - ANInstreamVideoAdLoadDelegate.

- (void)adDidReceiveAd:(id)ad
{
TESTTRACE();

    [self.expectationLoadVideoAd fulfill];
}

- (void)ad:(id)ad requestFailedWithError:(NSError *)error
{
TESTTRACE();
}


@end

File: tests/SwiftCompatibilityTest/SwiftCompatibilityTestTests/ANInstreamVideoAd/ANInstreamVideoAdTestCase.swift
/*
 *
 *    Copyright 2018 APPNEXUS INC
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import XCTest

class ANInstreamVideoAdTestCase: XCTestCase, ANInstreamVideoAdLoadDelegate {

    var instreamVideoAd: ANInstreamVideoAd!
    var expectationLoadVideoAd: XCTestExpectation!
    var request: URLRequest!
    var jsonRequestBody = [String : Any]()

    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.

        ANHTTPStubbingManager.shared().enable()
        ANHTTPStubbingManager.shared().ignoreUnstubbedRequests = true
        ANHTTPStubbingManager.shared().broadcastRequests = true
        NotificationCenter.default.addObserver(self, selector: #selector(self.requestCompleted(_:)), name: NSNotification.Name.anhttpStubURLProtocolRequestDidLoad, object: nil)
        request = nil
    }

    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
        instreamVideoAd = nil
        expectationLoadVideoAd = nil

        ANHTTPStubbingManager.shared().disable()
        ANHTTPStubbingManager.shared().removeAllStubs()
        ANHTTPStubbingManager.shared().broadcastRequests = false
        ANSDKSettings.sharedInstance().httpsEnabled = false
        NotificationCenter.default.removeObserver(self)
    }

    func requestCompleted(_ notification: Notification?) {
        var incomingRequest = notification?.userInfo![kANHTTPStubURLProtocolRequest] as? URLRequest
        let requestString = incomingRequest?.url?.absoluteString
        let searchString = ANSDKSettings.sharedInstance().baseUrlConfig.utAdRequestBaseUrl()
        if request == nil && requestString?.range(of:searchString!) != nil{
            request = notification!.userInfo![kANHTTPStubURLProtocolRequest] as? URLRequest
            jsonRequestBody = ANHTTPStubbingManager.jsonBodyOfURLRequest(asDictionary: request) as! [String : Any]
        }
    }

    // MARK: - Test methods.
    //Test video duration of InstreamVideoAd
    func test_TC32_AdDuration() {
        initializeInstreamVideoWithAllProperties()
        print("reached here")
        XCTAssertNotNil(instreamVideoAd)
        let duration = instreamVideoAd.getDuration()
        XCTAssertNotEqual(duration, 0)
    }

    //Test without setting the video duration of InstreamVideoAd
    func test_TC33_AdDurationNotSet() {
        initializeInstreamVideoWithNoProperties()
        XCTAssertNotNil(instreamVideoAd)
        let duration = instreamVideoAd.getDuration()
        XCTAssertEqual(duration, 0)
    }

    //Test vast url content of InstreamVideoAd
    func test_TC34_VastCreativeURL() {

        initializeInstreamVideoWithAllProperties()
        print("reached here")
        XCTAssertNotNil(instreamVideoAd)
        let vastcreativeTag = instreamVideoAd.getVastURL()
        XCTAssertNotNil(vastcreativeTag)
        XCTAssertNotEqual(vastcreativeTag?.count, 0)
        XCTAssertNotNil(vastcreativeTag)
        XCTAssertEqual(vastcreativeTag, "http://sampletag.com")
    }

    //Test without setting vast url content of InstreamVideoAd
    func test_TC35_VastCreativeValuesNotSet() {
        initializeInstreamVideoWithNoProperties()
        XCTAssertNotNil(instreamVideoAd)
        let vastcreativeTag = instreamVideoAd.getVastURL()
        XCTAssertEqual(vastcreativeTag?.count, 0)
    }

    //Test vast xml content of InstreamVideoAd
    func test_TC36_VastCreativeXML() {

        initializeInstreamVideoWithAllProperties()
        XCTAssertNotNil(instreamVideoAd)
        let vastcreativeXMLTag = instreamVideoAd.getVastXML()
        XCTAssertNotNil(vastcreativeXMLTag)
        XCTAssertNotEqual(vastcreativeXMLTag?.count, 0)
        XCTAssertNotNil(vastcreativeXMLTag)
        XCTAssertEqual(vastcreativeXMLTag, "http://sampletag.com")
    }

    //Test without setting the vast xml content of InstreamVideoAd
    func test_TC37_VastCreativeXMLValuesNotSet() {
        initializeInstreamVideoWithNoProperties()
        XCTAssertNotNil(instreamVideoAd)
        let vastcreativeXMLTag = instreamVideoAd.getVastXML()
        XCTAssertEqual(vastcreativeXMLTag?.count, 0)
    }

    //Test creative tag of InstreamVideoAd
    func test_TC38_CreativeTag() {
        initializeInstreamVideoWithAllProperties()
        XCTAssertNotNil(instreamVideoAd)
        let creativeTag = instreamVideoAd.getCreativeURL()
        XCTAssertNotEqual(creativeTag?.count, 0)
        XCTAssertNotNil(creativeTag)
        XCTAssertEqual(creativeTag, "http://sampletag.com")
    }

    //Test without setting the creative tag of InstreamVideoAd
    func test_TC39_CreativeValuesNotSet() {
        initializeInstreamVideoWithNoProperties()
        XCTAssertNotNil(instreamVideoAd)
        let creativeTag = instreamVideoAd.getCreativeURL()
        XCTAssertEqual(creativeTag?.count, 0)
    }

    //Test play head time for InstreamVideoAd
    func test_TC40_PlayHeadTimeForVideoSet() {
        initializeInstreamVideoWithNoProperties()
        XCTAssertNotNil(instreamVideoAd)
        let duration = instreamVideoAd.getPlayElapsedTime()
        XCTAssertNotEqual(duration, 0)
    }

    //Test creative tag of InstreamVideoAd
    func test_TC63_CustomKeywordsAdded() {
        instreamVideoAd = ANInstreamVideoAd(placementId: "12534678")
        instreamVideoAd.addCustomKeyword(withKey: "force_creative_id", value: "123456789")
        stubRequestWithResponse("SuccessfulInstreamVideoAdResponse")
        instreamVideoAd.load(with: self)
        expectationLoadVideoAd = expectation(description: "\(#function)")
        waitForExpectations(timeout: 30, handler: nil)
        XCTAssertEqual(instreamVideoAd.placementId, "12534678")
        if let arr = jsonRequestBody["keywords"] as? [Any], let dic = arr[0] as? [String : Any], let key = dic["key"] as? String
        {
            XCTAssertEqual(key, "force_creative_id")
        }
        if let arr = jsonRequestBody["keywords"] as? [Any], let dic = arr[0] as? [String : Any], let arr2 = dic["value"] as? [Any], let value = arr2[0] as? String
        {
            XCTAssertEqual(value, "123456789")
        }
    }

    // MARK: - Helper methods.
    func initializeInstreamVideoWithAllProperties() {
        instreamVideoAd = ANInstreamVideoAd()
        instreamVideoAd.adPlayer = ANVideoAdPlayer()
        instreamVideoAd.adPlayer.videoDuration = 10
        instreamVideoAd.adPlayer.creativeURL = "http://sampletag.com"
        instreamVideoAd.adPlayer.vastURLContent = "http://sampletag.com"
        instreamVideoAd.adPlayer.vastXMLContent = "http://sampletag.com"

    }

    func initializeInstreamVideoWithNoProperties() {
        instreamVideoAd = ANInstreamVideoAd()
        instreamVideoAd.adPlayer = ANVideoAdPlayer()
    }

    // MARK: - Stubbing
    func stubRequestWithResponse(_ responseName: String?) {
        let currentBundle = Bundle(for: type(of: self))
        let baseResponse = try? String(contentsOfFile: currentBundle.path(forResource: responseName, ofType: "json") ?? "", encoding: .utf8)
        let requestStub = ANURLConnectionStub()
        requestStub.requestURL = ANSDKSettings.sharedInstance().baseUrlConfig.utAdRequestBaseUrl()
        requestStub.responseCode = 200
        requestStub.responseBody = baseResponse

        ANHTTPStubbingManager.shared().add(requestStub)
    }

    @objc func fulfillExpectation(_ expectation: XCTestExpectation?) {
        expectation?.fulfill()
    }

    func waitForTimeInterval(_ delay: TimeInterval) {
        let expectation: XCTestExpectation = self.expectation(description: "wait")
        perform(#selector(self.fulfillExpectation(_:)), with: expectation, afterDelay: delay)

        waitForExpectations(timeout: TimeInterval(delay + 1), handler: nil)
    }

    // MARK: - ANInstreamVideoAdLoadDelegate.
    func adDidReceiveAd(_ ad: Any) {
        expectationLoadVideoAd.fulfill()
        expectationLoadVideoAd = nil;

    }
    func ad(_ ad: ANAdProtocol, requestFailedWithError error: Error) {
        print(error.localizedDescription)
    }
}
